odoo/websocket.py

```diff
@@ -0,0 +1,938 @@
+
+from odoo import api, evented, http
+from .http import root, Response, SessionExpiredException
+from .modules.registry import Registry
+from .service import model as service_model
+from .service.server import CommonServer
+
+_logger = logging.getLogger(__name__)
+
+
+class PollablePriorityQueue(queue.PriorityQueue):
+    """ A custom PriorityQueue than can be polled """
+    # This class allow this queue to be used with select.
+    def __init__(self):
+        super().__init__()
+        self._putsocket, self._getsocket = socket.socketpair()
+
+    def fileno(self):
+        return self._getsocket.fileno()
+
+    def put(self, item, **kwargs):
+        super().put(item, **kwargs)
+        self._putsocket.send(b'x')
+
+    def get(self, **kwargs):
+        self._getsocket.recv(1)
+        return super().get(**kwargs)
+
+
+# ------------------------------------------------------
+# EXCEPTIONS
+# ------------------------------------------------------
+
+class UpgradeRequired(HTTPException):
+    code = 426
+    description = "Wrong websocket version was given during the handshake"
+
+    def get_headers(self, environ=None):
+        headers = super().get_headers(environ)
+        headers.append((
+            'Sec-WebSocket-Version',
+            '; '.join(WebsocketConnectionHandler.SUPPORTED_VERSIONS)
+        ))
+        return headers
+
+
+class WebsocketException(Exception):
+    """ Base class for all websockets exceptions """
+
+
+class ConnectionClosed(WebsocketException):
+    """
+        Raised when the other end closes the socket without performing
+        the closing handshake.
+    """
+
+
+class InvalidCloseCodeException(WebsocketException):
+    def __init__(self, code):
+        super().__init__(f"Invalid close code: {code}")
+
+
+class InvalidDatabaseException(WebsocketException):
+    """
+        When raised: the database probably does not exists anymore, the
+        database is corrupted or the database version doesn't match the
+        server version.
+    """
+
+
+class InvalidWebsocketRequest(WebsocketException):
+    """
+        Raised when a websocket request is invalid (format, wrong args).
+    """
+
+
+class InvalidStateException(WebsocketException):
+    """
+        Raised when an operation is forbidden in the current state.
+    """
+
+
+class PayloadTooLargeException(WebsocketException):
+    """
+        Raised when a websocket message is too large.
+    """
+
+
+class ProtocolError(WebsocketException):
+    """
+        Raised when a frame format doesn't match expectations.
+    """
+
+
+class RateLimitExceededException(Exception):
+    """
+        Raised when a client exceeds the number of request in a given
+        time.
+    """
+
+
+# ------------------------------------------------------
+# RATE LIMITING
+# ------------------------------------------------------
+
+
+RL_DELAY = 0.2
+RL_BURST = 10
+_rate_limits_by_sid = defaultdict(functools.partial(deque, maxlen=RL_BURST))
+def limit_rate(sid):
```

Comment:
For sake of completeness, the deque could be an attribute and `limit_rate` a method of Websocket. The benefit is the automatic garbage collection when the websocket instance is not used anymore. This remove the need to `_rate_limites_by_sid.pop(sid)` upon disconnection.